HADOOP-16900. Very large files can be truncated when written through S3AFileSystem

steveloughran opened the pull request:
Contributed by Steve Loughran.

WriteOperationsHelper now raises a PathIOException when the number of
parts to write is too high.

This is cached during block uploads in S3ABlockOutputStream, so that once
one has been raised, there's no attempt to upload other blocks,
or to complete() the upload as if it was considered a success.

S3ABlockOutputStream also caches any failure raised during block write,
so things will fail earlier (currently it would only surface in close()),
and will abort() if anything went wrong during the upload.

This is intended to ensure that
* if too many parts are uploaded, the operation fails
* if anything causes a block upload to fail, the entire write is considered
a failure
* after a failure, the upload is never completed
* and abort() is always closed to remove any pending data

Change-Id: I7ee6cd1defcc4c7ccb1f95b933af2959cd14ddc2

hadoop-yetus commented:
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   0m 34s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  The patch appears to include 1 new or modified test files.  |
||| _ trunk Compile Tests _ |
| +1 :green_heart: |  mvninstall  |  19m  1s |  trunk passed  |
| +1 :green_heart: |  compile  |   0m 35s |  trunk passed  |
| +1 :green_heart: |  checkstyle  |   0m 27s |  trunk passed  |
| +1 :green_heart: |  mvnsite  |   0m 38s |  trunk passed  |
| +1 :green_heart: |  shadedclient  |  15m  4s |  branch has no errors when building and testing our client artifacts.  |
| +1 :green_heart: |  javadoc  |   0m 28s |  trunk passed  |
| +0 :ok: |  spotbugs  |   0m 58s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| +1 :green_heart: |  findbugs  |   0m 57s |  trunk passed  |
||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   0m 33s |  the patch passed  |
| +1 :green_heart: |  compile  |   0m 28s |  the patch passed  |
| +1 :green_heart: |  javac  |   0m 28s |  the patch passed  |
| -0 :warning: |  checkstyle  |   0m 18s |  hadoop-tools/hadoop-aws: The patch generated 2 new + 5 unchanged - 0 fixed = 7 total (was 5)  |
| +1 :green_heart: |  mvnsite  |   0m 31s |  the patch passed  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  shadedclient  |  14m  4s |  patch has no errors when building and testing our client artifacts.  |
| +1 :green_heart: |  javadoc  |   0m 24s |  the patch passed  |
| +1 :green_heart: |  findbugs  |   1m  1s |  the patch passed  |
||| _ Other Tests _ |
| +1 :green_heart: |  unit  |   1m 15s |  hadoop-aws in the patch passed.  |
| +1 :green_heart: |  asflicense  |   0m 32s |  The patch does not generate ASF License warnings.  |
|  |   |  58m 21s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.8 Server=19.03.8 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-1944/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/1944 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux 11a582320463 4.15.0-60-generic #67-Ubuntu SMP Thu Aug 22 16:55:30 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / bffb43b |
| Default Java | 1.8.0_242 |
| checkstyle | https://builds.apache.org/job/hadoop-multibranch/job/PR-1944/1/artifact/out/diff-checkstyle-hadoop-tools_hadoop-aws.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-1944/1/testReport/ |
| Max. process+thread count | 438 (vs. ulimit of 5500) |
| modules | C: hadoop-tools/hadoop-aws U: hadoop-tools/hadoop-aws |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-1944/1/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.11.1 https://yetus.apache.org |


This message was automatically generated.

steveloughran commented:
test failures

*  #1919
* timeout in one of the MR tests,  ITestS3ACommitterMRJob.test_200_execute:290 » , when running @ threads = 8. Seen that elsewhere.

I'm thinking about how best to test the cleanup logic.
* createFile() with block size == 1 (will also need to tweak create() to actually listen to block size)
* write 10K worth of blocks & expect failure
* verify that there are no pending uploads under the path
* and that metrics for abort ops > 0

steveloughran commented:
```
./hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3ABlockOutputStream.java:31:import com.amazonaws.AmazonClientException;:8: Unused import - com.amazonaws.AmazonClientException. [UnusedImports]
./hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3ABlockOutputStream.java:696:    /**: First sentence should end with a period. [JavadocStyle]
```

mukund-thakur commented on hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/commit/CommitOperations.java:
```diff
@@ -481,6 +483,15 @@ public SinglePendingCommit uploadFileToPendingCommit(File localFile,
       if (numParts == 0) {
         numParts = 1;
       }
+      if (numParts >= InternalConstants.UPLOAD_PART_COUNT_LIMIT) {
```
Should be greater than only right??

mukund-thakur commented on hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/WriteOperationHelper.java:
```diff
@@ -384,6 +387,7 @@ public void abortMultipartCommit(String destKey, String uploadId)
    * A subset of the file may be posted, by providing the starting point
    * in {@code offset} and a length of block in {@code size} equal to
    * or less than the remaining bytes.
+   * The part number must be less than 10000.
```
less than equal to 10000 ??

mukund-thakur replied:
??

mukund-thakur replied:
review this once.

mukund-thakur commented on hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/WriteOperationHelper.java:
```diff
@@ -400,18 +406,22 @@ public UploadPartRequest newUploadPartRequest(
 
     LOG.debug("Creating part upload request for {} #{} size {}",
         uploadId, partNumber, size);
+    if (partNumber >= InternalConstants.UPLOAD_PART_COUNT_LIMIT) {
```
Greater than??

mukund-thakur commented on hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/impl/InternalConstants.java:
```diff
@@ -93,4 +93,10 @@ private InternalConstants() {
+   * Maximum entries you can upload in a single file write/copy/upload.
+   * Value: {@value}.
+   */
+  public static final int UPLOAD_PART_COUNT_LIMIT = 10000;
```
Should mention here that this limit is from the AWS SDK side ?

mukund-thakur commented:
I think it is kind of impossible to write integration test for this as we have 5 MB minimum block size check at multiple places. Even if we remove those checks, we will have to do 10000 POST calls which will make the test super slow. 
Also, I tried removing the 5 MB checks but the test fails with StackOverflow error because of the tail recursion present at https://github.com/apache/hadoop/blob/5e0eda5d5f696aba7fc209874d232baf2a50d547/hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3ABlockOutputStream.java#L296

mukund-thakur commented on hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/WriteOperationHelper.java:
```diff
@@ -400,18 +406,22 @@ public UploadPartRequest newUploadPartRequest(
-        partNumber);
+    checkArgument(partNumber > 0,
+        "partNumber must be between 1 and %s inclusive, but is %s",
+        UPLOAD_PART_COUNT_LIMIT, partNumber);
```
The error message in argument check is not equivalent to argument condition.

mukund-thakur commented on hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3ABlockOutputStream.java:
```diff
@@ -394,6 +395,13 @@ public void close() throws IOException {
+      // operation, so as to not leave (billable) data
+      // pending on the bucket
+      if (multiPartUpload != null) {
+        multiPartUpload.abort();
```
I feel this abort call should already have been present? Does this mean we had a bug of having the partial data in S3 in case of other exceptions as well?

mukund-thakur commented:
> Also, I tried removing the 5 MB checks but the test fails with StackOverflow error because of the tail recursion present at

Here is the stack trace. 
Also the number bytes written is never zero. One of the possibility we were thinking and neither the recursion is going in an infinite loop. 
> java.lang.StackOverflowError
	at java.io.UnixFileSystem.getBooleanAttributes0(Native Method)
	at java.io.UnixFileSystem.getBooleanAttributes(UnixFileSystem.java:242)
	at java.io.File.isDirectory(File.java:849)
	at org.apache.hadoop.util.DiskChecker.checkAccessByFileMethods(DiskChecker.java:155)
	at org.apache.hadoop.util.DiskChecker.checkDirInternal(DiskChecker.java:100)
	at org.apache.hadoop.util.DiskChecker.checkDir(DiskChecker.java:77)
	at org.apache.hadoop.util.BasicDiskValidator.checkStatus(BasicDiskValidator.java:32)
	at org.apache.hadoop.fs.LocalDirAllocator$AllocatorPerContext.createPath(LocalDirAllocator.java:366)
	at org.apache.hadoop.fs.LocalDirAllocator$AllocatorPerContext.getLocalPathForWrite(LocalDirAllocator.java:446)
	at org.apache.hadoop.fs.LocalDirAllocator.getLocalPathForWrite(LocalDirAllocator.java:165)
	at org.apache.hadoop.fs.LocalDirAllocator.getLocalPathForWrite(LocalDirAllocator.java:146)
	at org.apache.hadoop.fs.s3a.S3AFileSystem.createTmpFileForWrite(S3AFileSystem.java:909)
	at org.apache.hadoop.fs.s3a.S3ADataBlocks$DiskBlockFactory.create(S3ADataBlocks.java:811)
	at org.apache.hadoop.fs.s3a.S3ABlockOutputStream.createBlockIfNeeded(S3ABlockOutputStream.java:192)
	at org.apache.hadoop.fs.s3a.S3ABlockOutputStream.write(S3ABlockOutputStream.java:286)
	at org.apache.hadoop.fs.s3a.S3ABlockOutputStream.write(S3ABlockOutputStream.java:301)
	at org.apache.hadoop.fs.s3a.S3ABlockOutputStream.write(S3ABlockOutputStream.java:301)
	at org.apache.hadoop.fs.s3a.S3ABlockOutputStream.write(S3ABlockOutputStream.java:301)

mukund-thakur commented:
Also the minimum upload size check is in the SDK side and it surfaces only when we issue finalizeMultipartUpload(). The reason we don't get size error during regular upload blocks. 
I am wondering isn't a bad design in terms for fail fast algorithm letting the clients wait unnecessarily knowing that it will fail eventually thus wasting time and compute both. Ideally S3 should have thrown error during the initial uploads only. 

> Caused by: com.amazonaws.services.s3.model.AmazonS3Exception: Your proposed upload is smaller than the minimum allowed size (Service: null; Status Code: 0; Error Code: EntityTooSmall; Request ID: CDD4C1CEE6A29607; S3 Extended Request ID: SR0kiTM3c9GAuv9lfXH+zq5cdj04gmk6/bKIYf8V73xugfi7257Cj51qy0YITShTJXGkK6Dw6AY=), S3 Extended Request ID: SR0kiTM3c9GAuv9lfXH+zq5cdj04gmk6/bKIYf8V73xugfi7257Cj51qy0YITShTJXGkK6Dw6AY=
	at com.amazonaws.services.s3.model.transform.XmlResponsesSaxParser$CompleteMultipartUploadHandler.doEndElement(XmlResponsesSaxParser.java:1815)
	at com.amazonaws.services.s3.model.transform.AbstractHandler.endElement(AbstractHandler.java:52)
	at com.sun.org.apache.xerces.internal.parsers.AbstractSAXParser.endElement(AbstractSAXParser.java:609)
	at com.sun.org.apache.xerces.internal.impl.XMLDocumentFragmentScannerImpl.scanEndElement(XMLDocumentFragmentScannerImpl.java:1782)
	at com.sun.org.apache.xerces.internal.impl.XMLDocumentFragmentScannerImpl$FragmentContentDriver.next(XMLDocumentFragmentScannerImpl.java:2967)
	at com.sun.org.apache.xerces.internal.impl.XMLDocumentScannerImpl.next(XMLDocumentScannerImpl.java:602)
	at com.sun.org.apache.xerces.internal.impl.XMLNSDocumentScannerImpl.next(XMLNSDocumentScannerImpl.java:112)
	at com.sun.org.apache.xerces.internal.impl.XMLDocumentFragmentScannerImpl.scanDocument(XMLDocumentFragmentScannerImpl.java:505)
	at com.sun.org.apache.xerces.internal.parsers.XML11Configuration.parse(XML11Configuration.java:842)
	at com.sun.org.apache.xerces.internal.parsers.XML11Configuration.parse(XML11Configuration.java:771)
	at com.sun.org.apache.xerces.internal.parsers.XMLParser.parse(XMLParser.java:141)
	at com.sun.org.apache.xerces.internal.parsers.AbstractSAXParser.parse(AbstractSAXParser.java:1213)
	at com.amazonaws.services.s3.model.transform.XmlResponsesSaxParser.parseXmlInputStream(XmlResponsesSaxParser.java:152)
	at com.amazonaws.services.s3.model.transform.XmlResponsesSaxParser.parseCompleteMultipartUploadResponse(XmlResponsesSaxParser.java:486)
	at com.amazonaws.services.s3.model.transform.Unmarshallers$CompleteMultipartUploadResultUnmarshaller.unmarshall(Unmarshallers.java:308)
	at com.amazonaws.services.s3.model.transform.Unmarshallers$CompleteMultipartUploadResultUnmarshaller.unmarshall(Unmarshallers.java:305)
	at com.amazonaws.services.s3.internal.S3XmlResponseHandler.handle(S3XmlResponseHandler.java:62)
	at com.amazonaws.services.s3.internal.ResponseHeaderHandlerChain.handle(ResponseHeaderHandlerChain.java:44)
	at com.amazonaws.services.s3.internal.ResponseHeaderHandlerChain.handle(ResponseHeaderHandlerChain.java:30)
	at com.amazonaws.http.response.AwsResponseHandlerAdapter.handle(AwsResponseHandlerAdapter.java:70)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.handleResponse(AmazonHttpClient.java:1627)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.executeOneRequest(AmazonHttpClient.java:1336)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.executeHelper(AmazonHttpClient.java:1113)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.doExecute(AmazonHttpClient.java:770)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.executeWithTimer(AmazonHttpClient.java:744)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.execute(AmazonHttpClient.java:726)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.access$500(AmazonHttpClient.java:686)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutionBuilderImpl.execute(AmazonHttpClient.java:668)
	at com.amazonaws.http.AmazonHttpClient.execute(AmazonHttpClient.java:532)
	at com.amazonaws.http.AmazonHttpClient.execute(AmazonHttpClient.java:512)
	at com.amazonaws.services.s3.AmazonS3Client.invoke(AmazonS3Client.java:4920)
	at com.amazonaws.services.s3.AmazonS3Client.invoke(AmazonS3Client.java:4866)
	at com.amazonaws.services.s3.AmazonS3Client.completeMultipartUpload(AmazonS3Client.java:3464)
	at org.apache.hadoop.fs.s3a.WriteOperationHelper.lambda$finalizeMultipartUpload$1(WriteOperationHelper.java:266)
	at org.apache.hadoop.fs.s3a.Invoker.once(Invoker.java:110)

mukund-thakur commented:
Looks like the stackoverflow error is completely normal (as in process is just running out of stack memory)  as it only happens when the number of parts is more that 4600. I couldn't find any reason where it may look like an infinite loop problem.

steveloughran commented:
> I am wondering isn't a bad design in terms for fail fast algorithm letting the clients wait unnecessarily knowing that it will fail eventually thus wasting time and compute both. Ideally S3 should have thrown error during the initial uploads only.

S3 doesn't know the order you actually want to complete the upload -any file could have been the last one. The order is defined in the final post just by the ordering of the etags.

steveloughran commented:
Test changes

* rename the test suite so future work on the MPU API doesnt conflict.
* split the two uploads into independent methods
* verify the oversize upload didn't create the file.
* change the limit property to fs.s3a.internal.upload.part.count.limit; add javadocs
* added a test for the committer code too, to verify exception is handled there too


Production
* range safety checks on loading that part count limit.
* from the initial failure of the committer test, discover and fix a double count of the commits aborted statistic

test: s3 ireland *new test only*

hadoop-yetus commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   1m 14s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  1s |  No case conflicting files found.  |
| +0 :ok: |  markdownlint  |   0m  0s |  markdownlint was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  The patch appears to include 4 new or modified test files.  |
||| _ trunk Compile Tests _ |
| +1 :green_heart: |  mvninstall  |  23m 37s |  trunk passed  |
| +1 :green_heart: |  compile  |   0m 31s |  trunk passed  |
| +1 :green_heart: |  checkstyle  |   0m 25s |  trunk passed  |
| +1 :green_heart: |  mvnsite  |   0m 36s |  trunk passed  |
| +1 :green_heart: |  shadedclient  |  17m 18s |  branch has no errors when building and testing our client artifacts.  |
| +1 :green_heart: |  javadoc  |   0m 27s |  trunk passed  |
| +0 :ok: |  spotbugs  |   1m  6s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| +1 :green_heart: |  findbugs  |   1m  4s |  trunk passed  |
||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   0m 37s |  the patch passed  |
| +1 :green_heart: |  compile  |   0m 30s |  the patch passed  |
| +1 :green_heart: |  javac  |   0m 30s |  the patch passed  |
| -0 :warning: |  checkstyle  |   0m 18s |  hadoop-tools/hadoop-aws: The patch generated 6 new + 5 unchanged - 0 fixed = 11 total (was 5)  |
| +1 :green_heart: |  mvnsite  |   0m 33s |  the patch passed  |
| -1 :x: |  whitespace  |   0m  0s |  The patch has 1 line(s) that end in whitespace. Use git apply --whitespace=fix <<patch_file>>. Refer https://git-scm.com/docs/git-apply  |
| -1 :x: |  whitespace  |   0m  0s |  The patch 5 line(s) with tabs.  |
| +1 :green_heart: |  shadedclient  |  15m 36s |  patch has no errors when building and testing our client artifacts.  |
| +1 :green_heart: |  javadoc  |   0m 23s |  the patch passed  |
| +1 :green_heart: |  findbugs  |   1m  1s |  the patch passed  |
||| _ Other Tests _ |
| +1 :green_heart: |  unit  |   1m 30s |  hadoop-aws in the patch passed.  |
| +1 :green_heart: |  asflicense  |   0m 28s |  The patch does not generate ASF License warnings.  |
|  |   |  67m 38s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.40 ServerAPI=1.40 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-1944/2/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/1944 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle markdownlint |
| uname | Linux 84c0799ef02b 4.15.0-74-generic #84-Ubuntu SMP Thu Dec 19 08:06:28 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / bd342bef64e |
| Default Java | Private Build-1.8.0_252-8u252-b09-1~18.04-b09 |
| checkstyle | https://builds.apache.org/job/hadoop-multibranch/job/PR-1944/2/artifact/out/diff-checkstyle-hadoop-tools_hadoop-aws.txt |
| whitespace | https://builds.apache.org/job/hadoop-multibranch/job/PR-1944/2/artifact/out/whitespace-eol.txt |
| whitespace | https://builds.apache.org/job/hadoop-multibranch/job/PR-1944/2/artifact/out/whitespace-tabs.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-1944/2/testReport/ |
| Max. process+thread count | 338 (vs. ulimit of 5500) |
| modules | C: hadoop-tools/hadoop-aws U: hadoop-tools/hadoop-aws |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-1944/2/console |
| versions | git=2.17.1 maven=3.6.0 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

mukund-thakur commented on hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/ITestS3ABlocksize.java:
```diff
@@ -18,9 +18,12 @@
 
 package org.apache.hadoop.fs.s3a;
 
+import java.io.IOException;
```
remove.

mukund-thakur commented on hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/ITestS3ABlocksize.java:
```diff
@@ -18,9 +18,12 @@
 import org.apache.hadoop.fs.FileStatus;
 import org.apache.hadoop.fs.FileSystem;
 import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.fs.s3a.impl.InternalConstants;
```
remove

steveloughran commented:
ok i'll get this through checkstyle and merge in

hadoop-yetus commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |  22m 48s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +0 :ok: |  markdownlint  |   0m  1s |  markdownlint was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  The patch appears to include 4 new or modified test files.  |
||| _ trunk Compile Tests _ |
| +1 :green_heart: |  mvninstall  |  20m 13s |  trunk passed  |
| +1 :green_heart: |  compile  |   0m 37s |  trunk passed  |
| +1 :green_heart: |  checkstyle  |   0m 27s |  trunk passed  |
| +1 :green_heart: |  mvnsite  |   0m 40s |  trunk passed  |
| +1 :green_heart: |  shadedclient  |  15m  7s |  branch has no errors when building and testing our client artifacts.  |
| +1 :green_heart: |  javadoc  |   0m 30s |  trunk passed  |
| +0 :ok: |  spotbugs  |   1m  0s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| +1 :green_heart: |  findbugs  |   0m 57s |  trunk passed  |
||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   0m 33s |  the patch passed  |
| +1 :green_heart: |  compile  |   0m 28s |  the patch passed  |
| +1 :green_heart: |  javac  |   0m 28s |  the patch passed  |
| -0 :warning: |  checkstyle  |   0m 19s |  hadoop-tools/hadoop-aws: The patch generated 2 new + 5 unchanged - 0 fixed = 7 total (was 5)  |
| +1 :green_heart: |  mvnsite  |   0m 32s |  the patch passed  |
| -1 :x: |  whitespace  |   0m  0s |  The patch has 1 line(s) that end in whitespace. Use git apply --whitespace=fix <<patch_file>>. Refer https://git-scm.com/docs/git-apply  |
| -1 :x: |  whitespace  |   0m  0s |  The patch 5 line(s) with tabs.  |
| +1 :green_heart: |  shadedclient  |  13m 49s |  patch has no errors when building and testing our client artifacts.  |
| +1 :green_heart: |  javadoc  |   0m 26s |  the patch passed  |
| +1 :green_heart: |  findbugs  |   1m  2s |  the patch passed  |
||| _ Other Tests _ |
| +1 :green_heart: |  unit  |   1m 27s |  hadoop-aws in the patch passed.  |
| +1 :green_heart: |  asflicense  |   0m 31s |  The patch does not generate ASF License warnings.  |
|  |   |  81m 52s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.40 ServerAPI=1.40 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-1944/3/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/1944 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle markdownlint |
| uname | Linux d957cfc1601c 4.15.0-60-generic #67-Ubuntu SMP Thu Aug 22 16:55:30 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / d4e36409d40 |
| Default Java | Private Build-1.8.0_252-8u252-b09-1~18.04-b09 |
| checkstyle | https://builds.apache.org/job/hadoop-multibranch/job/PR-1944/3/artifact/out/diff-checkstyle-hadoop-tools_hadoop-aws.txt |
| whitespace | https://builds.apache.org/job/hadoop-multibranch/job/PR-1944/3/artifact/out/whitespace-eol.txt |
| whitespace | https://builds.apache.org/job/hadoop-multibranch/job/PR-1944/3/artifact/out/whitespace-tabs.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-1944/3/testReport/ |
| Max. process+thread count | 401 (vs. ulimit of 5500) |
| modules | C: hadoop-tools/hadoop-aws U: hadoop-tools/hadoop-aws |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-1944/3/console |
| versions | git=2.17.1 maven=3.6.0 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

steveloughran commented:
merged in. thanks for your work @mukund-thakur
